pkg/resources/file_format.go

```diff
@@ -0,0 +1,1018 @@
+	},
+	"preserve_space": {
+		Type:        schema.TypeBool,
+		Optional:    true,
+		Description: "Boolean that specifies whether the XML parser preserves leading and trailing spaces in element content.",
+	},
+	"strip_outer_element": {
+		Type:        schema.TypeBool,
+		Optional:    true,
+		Description: "Boolean that specifies whether the XML parser strips out the outer XML element, exposing 2nd level elements as separate documents.",
+	},
+	"disable_snowflake_data": {
+		Type:        schema.TypeBool,
+		Optional:    true,
+		Description: "Boolean that specifies whether the XML parser disables recognition of Snowflake semi-structured data tags.",
+	},
+	"disable_auto_convert": {
+		Type:        schema.TypeBool,
+		Optional:    true,
+		Description: "Boolean that specifies whether the XML parser disables automatic conversion of numeric and Boolean values from text to native representation.",
+	},
+	"comment": {
+		Type:        schema.TypeString,
+		Optional:    true,
+		Description: "Specifies a comment for the file format.",
+	},
+}
+
+type fileFormatID struct {
+	DatabaseName   string
+	SchemaName     string
+	FileFormatName string
+}
+
+// String() takes in a fileFormatID object and returns a pipe-delimited string:
+// DatabaseName|SchemaName|FileFormatName
+func (fi *fileFormatID) String() (string, error) {
+	var buf bytes.Buffer
+	csvWriter := csv.NewWriter(&buf)
+	csvWriter.Comma = fileFormatIDDelimiter
+	dataIdentifiers := [][]string{{fi.DatabaseName, fi.SchemaName, fi.FileFormatName}}
+	err := csvWriter.WriteAll(dataIdentifiers)
+	if err != nil {
+		return "", err
+	}
+	strFileFormatID := strings.TrimSpace(buf.String())
+	return strFileFormatID, nil
+}
+
+// fileFormatIDFromString() takes in a pipe-delimited string: DatabaseName|SchemaName|FileFormatName
+// and returns a fileFormatID object
+func fileFormatIDFromString(stringID string) (*fileFormatID, error) {
+	reader := csv.NewReader(strings.NewReader(stringID))
+	reader.Comma = fileFormatIDDelimiter
+	lines, err := reader.ReadAll()
+	if err != nil {
+		return nil, fmt.Errorf("Not CSV compatible")
+	}
+
+	if len(lines) != 1 {
+		return nil, fmt.Errorf("1 line per file format")
+	}
+	if len(lines[0]) != 3 {
+		return nil, fmt.Errorf("3 fields allowed")
+	}
+
+	fileFormatResult := &fileFormatID{
+		DatabaseName:   lines[0][0],
+		SchemaName:     lines[0][1],
+		FileFormatName: lines[0][2],
+	}
+	return fileFormatResult, nil
+}
+
+// FileFormat returns a pointer to the resource representing a file format
+func FileFormat() *schema.Resource {
+	return &schema.Resource{
+		Create: CreateFileFormat,
+		Read:   ReadFileFormat,
+		Update: UpdateFileFormat,
+		Delete: DeleteFileFormat,
+		Exists: FileFormatExists,
+
+		Schema: fileFormatSchema,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+	}
+}
+
+// CreateFileFormat implements schema.CreateFunc
+func CreateFileFormat(data *schema.ResourceData, meta interface{}) error {
```

Comment:
Did you consider providing some validation on field / parameter compatibility? Not all format types support all parameters. As it stands, if a user were to use an incompatible combination, they would get an sql error from snowflake, which might be confusing.

Reply:
Just added some validation to the Create method.

Reply:
The validation gets messier though since each Format Type has their own defaults too. Thinking they should each be separate resources to make it easier to take advantage of `ValidateFunc`. Thoughts?

Reply:
Unfortunately there is no good place to do cross-attribute validations except for in the Create and Update functions. ValidateFunc works at the scope of a single attribute.

I looked at two paths of tackling this awhile ago-

1. create a separate resource for each file format type. Probably ok since you can't change the format type anyway. Would add a lot of resources with a fair amount of duplication of code.

2. rather than take flat approach (which you are doing), I tried creating a structured type for each format. My WIP code is in https://github.com/chanzuckerberg/terraform-provider-snowflake/tree/ryanking/file_format. I gave up because I was trying to do something too hard - using the presence of and structure to signal the type (even if empty). Terraform doesn't do well with empty / zero stuff.

I think maybe a hybrid approach of 2 and what you are doing might work - create a separate structured attribute for each type, and also have a type attribute. Then we get some deal of validation via the type system.

Make sense?

Reply:
Yep, I think that hybrid approach makes sense. I'll see if I can sort it out.

Reply:
@ryanking I'll be picking this work up from @ajbosco in the coming days. Would (1) above be acceptable as an approach to this problem? I see how lots of resources would be added, but the approach is more straightforward (if verbose) to implement.

Reply:
I think we should either take option 2 or revert to @ajbosco's original approach of a flat list of options.